cmd/ingest/ingest.go

```diff
@@ -187,50 +393,82 @@ func (ing *ingester) sendResolvedURLs1(e WhitelistEntity, mustBeCharm bool, c ch
 		// Go through the published channels, finding out if any of them
 		// are mentioned on the requested channels. If so, we'll include the
 		// entity in that channel.
-		for pch, current := range result.published {
-			log.Printf("id %v; chan %v; published chan %v; current %v", curl, ch, pch, current)
+		for pch := range result.channels {
 			if !needChannels[pch] {
+				delete(result.channels, pch)
 				continue
 			}
-			c <- resolvedURL{
-				id:            result.id,
-				promulgatedId: result.promulgatedId,
-				channel:       pch,
-				// Note that we never mark anything current when a specific
-				// revision is mentioned.
-				current: current && curl.Revision == -1,
+			if curl.Revision != -1 {
+				// We only release a charm as the current version for a channel
+				// when the revision hasn't been explicitly specified.
+				result.channels[pch] = false
 			}
-			if result.id.Series == "bundle" {
-				if mustBeCharm {
-					return errgo.Newf("charm URL in bundle refers to bundle (%q) not charm", curl)
-				}
-				if result.bundleData == nil {
-					return errgo.Newf("bundle %q has no metadata", curl)
-				}
-				ing.sendResolvedURLsForBundle(curl, result.bundleData, c)
+		}
+		c <- result
+		if result.id.Series == "bundle" {
+			if mustBeCharm {
+				return errgo.Newf("charm URL in bundle refers to bundle (%q) not charm", curl)
```

Comment:
what are the chances of needing this? would charmstore allow a bundle with another embedded bundle to be uploaded? if it will then why not copy it anyway as it wouldn't seem to cause problems upstream.